File: frontend/src/views/Connection/DataView/components/PublishPanel/stores/publish-history.ts
import {
  GetPublishHistoriesForConnection,
  SavePublishHistoryEntry,
  DeletePublishHistoryEntry,
} from "bindings/mqtt-viewer/backend/app/app";
import { writable } from "svelte/store";
import * as app from "bindings/mqtt-viewer/backend/app/models";
import * as models from "bindings/mqtt-viewer/backend/models/models";

import type { DeepOmit } from "@/util/types";
import {
  publishDetailsFromStoredMessage,
  type PublishDetailsStore,
} from "./publish-details";

export type PublishHistory = DeepOmit<models.PublishHistory, "convertValues">[];

interface PublishHistoryStore {
  publishHistory: PublishHistory;
}

export const createPublishHistoryStore = (
  connId: number,
  publishDetailsStore: PublishDetailsStore
) => {
  const { subscribe, set, update } = writable<PublishHistoryStore>(
    {
      publishHistory: [],
    },
    (set) => {
      loadConnectionPublishHistory();
    }
  );

  const loadConnectionPublishHistory = async () => {
    try {
      const publishHistories = await GetPublishHistoriesForConnection(connId);
      if (publishHistories.length === 0) {
        return;
      }
      // @ts-ignore
      set({ publishHistory: publishHistories });
    } catch (e) {
      console.error(e);
    }
  };

  const setPublishDetailsFromHistoryEntry = (entry: models.PublishHistory) => {
    publishDetailsStore.setPartial({
      ...publishDetailsFromStoredMessage(entry),
      connectionId: connId,
      hasAttemptedPublish: true,
    });
  };

  const savePublishEntry = async (params: {
    connectionId: number;
    topic: string;
    payload: string;
    qos: number;
    retain: boolean;
    encoding: string;
    format: string;
    properties?: app.PublishProperties;
    userProperties?: { [key: string]: string };
    protoOverride?: string | null;
  }) => {
    try {
      console.log("saving publish entry", params);
      let userPropertiesString: undefined | string = undefined;
      if (!!params.userProperties) {
        try {
          userPropertiesString = JSON.stringify(params.userProperties);
        } catch (e) {
          console.error(e);
        }
      }
      const entry = await SavePublishHistoryEntry({
        connectionId: params.connectionId,
        topic: params.topic,
        payload: params.payload,
        qos: params.qos,
        retain: params.retain,
        encoding: params.encoding,
        format: params.format,
        userProperties: userPropertiesString ?? null,
        headerContentType: params.properties?.contentType ?? null,
        headerPayloadFormatIndicator:
          params.properties?.payloadFormatIndicator ?? null,
        headerMessageExpiryInterval:
          params.properties?.messageExpiryInterval ?? null,
        headerResponseTopic: params.properties?.responseTopic ?? null,
        headerCorrelationData: params.properties?.correlationData ?? null,
        headerSubscriptionIdentifier:
          params.properties?.subscriptionIdentifier ?? null,
        headerTopicAlias: params.properties?.topicAlias ?? null,
        protoOverride: params.protoOverride ?? null,
      });
      console.log("saved publish history entry", entry);
      update((store) => {
        //@ts-ignore - issue with omitting userProperties
        store.publishHistory = [entry, ...store.publishHistory];
        return store;
      });
    } catch (e) {
      console.error(e);
    }
  };

  const deletePublishEntry = async (id: number) => {
    try {
      await DeletePublishHistoryEntry(id);
      update((store) => {
        store.publishHistory = store.publishHistory.filter((e) => e.id !== id);
        return store;
      });
    } catch (e) {
      throw e;
    }
  };

  return {
    subscribe,
    savePublishEntry,
    deletePublishEntry,
    setPublishDetailsFromHistoryEntry,
  };
};

File: backend/app/publish_history.go
package app

import (
	"mqtt-viewer/backend/models"
	"time"
)

func (a *App) GetPublishHistoriesForConnection(connectionID uint) ([]models.PublishHistory, error) {
	var publishHistories []models.PublishHistory
	if err := a.Db.
		Where("connection_id = ?", connectionID).
		Order("published_at desc").
		Limit(500).
		Find(&publishHistories).Error; err != nil {
		return nil, err
	}
	return publishHistories, nil
}

type SavePublishHistoryEntryParams struct {
	ConnectionId                 uint    `json:"connectionId"`
	Topic                        string  `json:"topic"`
	Payload                      string  `json:"payload"`
	QoS                          uint    `json:"qos"`
	Retain                       bool    `json:"retain"`
	Encoding                     string  `json:"encoding"`
	Format                       string  `json:"format"`
	HeaderContentType            *string `json:"headerContentType"`
	HeaderResponseTopic          *string `json:"headerResponseTopic"`
	HeaderCorrelationData        *string `json:"headerCorrelationData"`
	HeaderPayloadFormatIndicator *bool   `json:"headerPayloadFormatIndicator"`
	HeaderMessageExpiryInterval  *int32  `json:"headerMessageExpiryInterval"`
	HeaderTopicAlias             *int32  `json:"headerTopicAlias"`
	HeaderSubscriptionIdentifier *int32  `json:"headerSubscriptionIdentifier"`
	UserPropertiesString         *string `json:"userProperties"`
}

func (a *App) SavePublishHistoryEntry(params SavePublishHistoryEntryParams) (models.PublishHistory, error) {
	entry := models.PublishHistory{
		ConnectionID: params.ConnectionId,
		Topic:        params.Topic,
		Payload:      params.Payload,
		QoS:          params.QoS,
		Retain:       params.Retain,
		Encoding:     params.Encoding,
		Format:       params.Format,
		PublishedAt:  time.Now(),
	}
	if params.UserPropertiesString != nil {
		entry.UserProperties = params.UserPropertiesString
	}
	if params.HeaderContentType != nil {
		entry.HeaderContentType = params.HeaderContentType
	}
	if params.HeaderResponseTopic != nil {
		entry.HeaderResponseTopic = params.HeaderResponseTopic
	}
	if params.HeaderCorrelationData != nil {
		entry.HeaderCorrelationData = params.HeaderCorrelationData
	}
	if params.HeaderPayloadFormatIndicator != nil {
		entry.HeaderPayloadFormatIndicator = params.HeaderPayloadFormatIndicator
	}
	if params.HeaderMessageExpiryInterval != nil {
		entry.HeaderMessageExpiryInterval = params.HeaderMessageExpiryInterval
	}
	if params.HeaderTopicAlias != nil {
		entry.HeaderTopicAlias = params.HeaderTopicAlias
	}
	if params.HeaderSubscriptionIdentifier != nil {
		entry.HeaderSubscriptionIdentifier = params.HeaderSubscriptionIdentifier
	}

	result := a.Db.Create(&entry)
	if result.Error != nil {
		return models.PublishHistory{}, result.Error
	}
	return entry, nil
}

func (a *App) DeletePublishHistoryEntry(id uint) error {
	err := a.Db.Where("id = ?", id).Delete(&models.PublishHistory{}).Error
	if err != nil {
		return err
	}
	return nil
}
